### ISSUE TITLE
compilation issue for plugins using coq_makefile

### ISSUE THREAD

**Issue by Zimmi48** [Zimmi48/bugzilla-test#5332]
Note: the issue was created automatically with bugzilla2github tool

# Bugzilla Bug ID: 5332
Date: 2017-01-31 16:36:00 +0100
From: simon.cruanes@inria.fr
To:   
CC:   coq-bugs-redist@lists.gforge.inria.fr, @_ejgallego, @_gares

Last updated: 2017-02-01 17:18:43 +0100

## Bugzilla Comment ID: 15247
Date: 2017-01-31 16:36:00 +0100
From: simon.cruanes@inria.fr

I am trying to write a plugin, using coq_makefile. It was primitive but would compile and load properly. However, I tried to switch to Coq 8.6, updating the .ml code but not touching _CoqProject, and now compilation fails as follows.
I really have no idea how to fix this.

$ make
CAMLDEP src/nunchaku_coq_sexp.mli
CAMLDEP src/nunchaku_coq_utils.mli
COQDEP src/nunchaku_coq.mllib
*** Warning: nunchaku_coq.mllib already found in src/ (discarding src/nunchaku_coq.mllib)
CAMLDEP src/nunchaku_coq_run.ml
CAMLDEP src/nunchaku_coq_sexp_lex.ml
CAMLDEP src/nunchaku_coq_sexp.ml
CAMLDEP src/nunchaku_coq_utils.ml
CAMLDEP src/nunchaku_coq_ast.ml
CAMLDEP -pp src/nunchaku_coq_main.ml4
COQDEP src/Nunchaku.v
CAMLC -c src/nunchaku_coq_utils.mli
CAMLC -c src/nunchaku_coq_utils.ml
CAMLC -c src/nunchaku_coq_ast.ml
CAMLC -c src/nunchaku_coq_sexp_lex.ml
CAMLC -c src/nunchaku_coq_sexp.mli
CAMLC -c src/nunchaku_coq_sexp.ml
CAMLC -c src/nunchaku_coq_run.ml
CAMLC -pp -c src/nunchaku_coq_main.ml4
CAMLC -a -o src/nunchaku_coq.cma
CAMLOPT -c src/nunchaku_coq_utils.ml
File "_none_", line 1:
Warning 58: no cmx file was found in path for module Term, and its interface was not compiled with -opaque
File "_none_", line 1:
Warning 58: no cmx file was found in path for module Names, and its interface was not compiled with -opaque
CAMLOPT -c src/nunchaku_coq_ast.ml
File "_none_", line 1:
Warning 58: no cmx file was found in path for module Names, and its interface was not compiled with -opaque
CAMLOPT -c src/nunchaku_coq_sexp_lex.ml
CAMLOPT -c src/nunchaku_coq_sexp.ml
CAMLOPT -c src/nunchaku_coq_run.ml
File "_none_", line 1:
Warning 58: no cmx file was found in path for module Declarations, and its interface was not compiled with -opaque
File "_none_", line 1:
Warning 58: no cmx file was found in path for module Pp, and its interface was not compiled with -opaque
File "_none_", line 1:
Warning 58: no cmx file was found in path for module Coqlib, and its interface was not compiled with -opaque
File "_none_", line 1:
Warning 58: no cmx file was found in path for module Names, and its interface was not compiled with -opaque
File "_none_", line 1:
Warning 58: no cmx file was found in path for module Globnames, and its interface was not compiled with -opaque
File "_none_", line 1:
Warning 58: no cmx file was found in path for module Constr, and its interface was not compiled with -opaque
File "_none_", line 1:
Warning 58: no cmx file was found in path for module Termops, and its interface was not compiled with -opaque
File "_none_", line 1:
Warning 58: no cmx file was found in path for module Environ, and its interface was not compiled with -opaque
File "_none_", line 1:
Warning 58: no cmx file was found in path for module Proofview, and its interface was not compiled with -opaque
File "_none_", line 1:
Warning 58: no cmx file was found in path for module Util, and its interface was not compiled with -opaque
File "_none_", line 1:
Warning 58: no cmx file was found in path for module Tacticals, and its interface was not compiled with -opaque
CAMLOPT -pp -c src/nunchaku_coq_main.ml4
File "_none_", line 1:
Warning 58: no cmx file was found in path for module Mltop, and its interface was not compiled with -opaque
File "_none_", line 1:
Warning 58: no cmx file was found in path for module Tacenv, and its interface was not compiled with -opaque
File "_none_", line 1:
Warning 58: no cmx file was found in path for module Names, and its interface was not compiled with -opaque
File "_none_", line 1:
Warning 58: no cmx file was found in path for module Loc, and its interface was not compiled with -opaque
File "_none_", line 1:
Warning 58: no cmx file was found in path for module Tacentries, and its interface was not compiled with -opaque
CAMLOPT -a -o src/nunchaku_coq.cmxa
CAMLOPT -shared -o src/nunchaku_coq.cmxs
COQC src/Nunchaku.v
File "./src/Nunchaku.v", line 4, characters 0-33:
Error: while loading
/home/simon/workspace/nunchaku-coq/src/nunchaku_coq.cmxs:
no implementation available for Declarations
make: *** [Makefile:496: src/Nunchaku.vo] Error 1

So it seems that the .cmxs doesn't embed some modules?

The project code can be found at:
https://github.com/nunchaku-inria/nunchaku-coq/tree/e1c0fe4bdfeb34becec961cca3a3370893748f4a

I use OCaml 4.03 and coq 8.6.

## Bugzilla Comment ID: 15249
Date: 2017-01-31 19:45:32 +0100
From: @_ejgallego

Yeah this looks like the typical Ocaml native linker not finding its path problem... I got bit by this but I forgot was the workaround was.

> make byte

should works for your case until more knowledgeable people can help.

## Bugzilla Comment ID: 15252
Date: 2017-02-01 09:29:30 +0100
From: simon.cruanes@inria.fr

sadly I get a very similar error with `make byte`:

COQC src/Nunchaku.v
File "./src/Nunchaku.v", line 4, characters 0-33:
Error:
Cannot link ml-object nunchaku_coq.cma to Coq code (reference to undefined global Declarations).

## Bugzilla Comment ID: 15253
Date: 2017-02-01 13:35:11 +0100
From: @_ejgallego

Make byte works for me with Coq 8.6 + Ocaml 4.04.

## Bugzilla Comment ID: 15255
Date: 2017-02-01 15:06:24 +0100
From: simon.cruanes@inria.fr

I tried it (both on 4.03 and 4.04.0) and it seems that `make byte` also compiles the native code, for some reason‽ Strangely, on 4.04.0, `make byte` succeeds but loading the module in Coq afterwards fails because it still tries to load the .cmxs.

## Bugzilla Comment ID: 15256
Date: 2017-02-01 17:10:23 +0100
From: @_gares

Hum, things are a bit unfortunate.

The coq sources include a file called checker/declarations.ml
Such file is in a path coq_makefile passes via -I to ocamlc, so your code compiles
The checker/ directory is not part of the coq binary. It contains a second kernel, a standalone checker for proof objects.
The coq binary is only built using declarations.mli (that is in kernel/) and only includes datatype declarations.

Your code does "module D = Declarations".  Such statement changed slightly 
"implementation" in OCaml 4.04, so this explains the difference you see.  In older ocaml versions it was provoking the duplication of the module code, while now it is just an alias.

In short, remove the "module D = ..." and use "Declarations".

It works here.  I've no time to check why it was working in 8.5, but probably we had a useless declarations.ml file in kernel/ too.

Best,

## Bugzilla Comment ID: 15257
Date: 2017-02-01 17:18:43 +0100
From: simon.cruanes@inria.fr

Interesting, I see now. It works, so thanks!
